fix(lora): add run_lora_a_embedding to ChunkedSgmvLoRABackend

ashtonchew opened the pull request:
## Motivation

Fixes #14773. This PR is a follow-up to #14796, which introduced a workaround by excluding `embed_tokens` and `lm_head` from `--lora-target-modules all` when using the `csgmv` backend.

Rather than disabling this functionality, this PR implements the missing `run_lora_a_embedding()` method in `ChunkedSgmvLoRABackend`, enabling full embedding LoRA support for the chunked SGMV backend. This addresses the TODO comment added in #14796.

## Modifications

- Added `embedding_lora_a_fwd` import from triton_ops
- Added `run_lora_a_embedding()` method using existing triton kernel
- Added `cuda_graph_embedding_batch_info` for CUDA graph mode
- Added `embedding_batch_info` setup in `prepare_lora_batch()` to maintain original sequence structure (not chunked/reordered like linear layers)
- Added unit test `test/srt/lora/test_chunked_lora_embedding.py`

## Test Plan

- [x] Added unit test `test/srt/lora/test_chunked_lora_embedding.py` that verifies:
  - `embedding_batch_info` is created with original sequence structure
  - `permutation=None` (no reordering for embeddings)
  - Output shape is correct `(seq_len, rank)`
  - Output matches direct kernel call with `embedding_batch_info`

## Accuracy Tests

N/A - This fix adds missing functionality that was causing a crash. It does not change existing model outputs; it enables a previously broken code path to work correctly.

## Benchmarking and Profiling

N/A - This fix adds missing functionality. The implementation follows the same pattern as `TritonLoRABackend.run_lora_a_embedding()` and reuses existing optimized triton kernels, so no performance regression is expected.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). *(N/A - Bug fix, no user-facing documentation changes needed)*
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). *(N/A - Bug fix that enables previously crashing functionality)*
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

ashtonchew committed:
fix(lora): add run_lora_a_embedding to ChunkedSgmvLoRABackend

Fixes #14773. When using --lora-target-modules all with the csgmv backend,
the embedding layer LoRA was failing because ChunkedSgmvLoRABackend did not
implement run_lora_a_embedding().

This commit adds:
- run_lora_a_embedding() method using existing triton kernel
- Separate embedding_batch_info for original sequence structure
- CUDA graph support for embedding LoRA
- Unit test for the new method

ashtonchew committed:
Merge branch 'main' into fix/lora-embedding-chunked-backend

ashtonchew committed:
Merge branch 'main' into fix/lora-embedding-chunked-backend

ashtonchew committed:
Merge branch 'main' into fix/lora-embedding-chunked-backend

ashtonchew committed:
Merge branch 'main' into fix/lora-embedding-chunked-backend

ashtonchew committed:
Merge branch 'main' into fix/lora-embedding-chunked-backend

ashtonchew commented:
@yushengsu-thu 
Sounds good, thanks for the clarification, I will fix this by tonight with your suggestion.

ashtonchew committed:
refactor(lora): replace random-init script with training script

- Delete create_embedding_lora_adapter.py (random weights)
- Add train_embedding_lora_adapter.py (actual fine-tuning)
- Relax logprob threshold for embedding LoRA test

ashtonchew committed:
Merge branch 'main' into fix/lora-embedding-chunked-backend

ashtonchew committed:
test(lora): add triton backend test for embedding LoRA

Add test_lora_embedding_logprob_comparison_triton to validate triton
backend as baseline before testing chunked SGMV backend.

yushengsu-thu reviewed (commented):
@ashtonchew 
In these two test scripts:
[test/srt/lora/test_lora_eviction.py](https://github.com/sgl-project/sglang/pull/14177/files#diff-90023cbea116a8e77c2cc1b665e0aff868b7e0bef028208f9b3cf0561b556741) and [test/srt/lora/test_lora_update.py](https://github.com/sgl-project/sglang/pull/14177/files#diff-0b03ad57c07397776998ae718b6e69f761f080e0b89aa802b8b1dc7e4c7d407e),
could you also add the same test cases as the original ones in these two files, and set `lora_target_modules=["all"]` along with use your trained lora weight. 

Also, could you comment out this part of the code (https://github.com/sgl-project/sglang/pull/14796/files) to verify that the necessary LoRA CI/CD run with `ChunkedSgmvLoRABackend` and checks still pass on your side?
Feel free to ping me if you get any questions.

ashtonchew committed:
test(lora): add embedding LoRA tests with lora_target_modules=all for csgmv

- Comment out csgmv embed_tokens/lm_head restriction in server_args.py
  since csgmv backend now supports these via run_lora_a_embedding()
- Add test_lora_eviction_with_embedding_lora_all_target_modules test
- Add EMBEDDING_LORA_TESTS to test_lora_update.py with 2 test cases
- Uses ash256/sglang_embedding_lora_test_adapter for testing

ashtonchew commented:
@yushengsu-thu  I verified that LoRA CI/CD passes with ChunkedSgmvLoRABackend.

## Changes Made

1. **Commented out the csgmv embed_tokens/lm_head restriction** in `python/sglang/srt/server_args.py`:
```python
# NOTE: The following code has been commented out because the csgmv backend
# now supports embedding and lm_head layers via run_lora_a_embedding().
# See: https://github.com/sgl-project/sglang/pull/14796
#
# if self.lora_backend == "csgmv":
#     logger.warning(...)
#     self.lora_target_modules.discard("embed_tokens")
#     self.lora_target_modules.discard("lm_head")
```

2. **Added test cases** with `lora_target_modules=["all"]` using `ash256/sglang_embedding_lora_test_adapter`:
   - `test/srt/lora/test_lora_eviction.py`: Added `test_lora_eviction_with_embedding_lora_all_target_modules`
   - `test/srt/lora/test_lora_update.py`: Added `EMBEDDING_LORA_TESTS` (2 test cases)

## Test Results

### Test 1: `test_lora_eviction_with_embedding_lora_all_target_modules`
```
========== Testing embedding LoRA eviction, repeat 1/2 ==========
prompt: AI is a field of computer science focused on...
output: developing artificial intelligence systems that can be used to solve complex problems. It is a rapid...

========== Testing embedding LoRA eviction, repeat 2/2 ==========
prompt: AI is a field of computer science focused on...
output: developing artificial intelligence systems that can be used to solve complex problems. It is a rapid...
PASSED

1 passed, 3 warnings in 56.98s
```

### Test 2: Embedding LoRA with `lora_target_modules=["all"]` on csgmv backend
```
================================================================================
Running EMBEDDING_LORA_TESTS with lora_target_modules=['all']
================================================================================

Test 1: Forward pass with embedding LoRA adapter
Output: [' great way to learn about the history of the world...', ' developing artificial intelligence systems that can be used to solve complex problems...']

Test 2: Forward pass with base model (no LoRA)
Output: [' leading provider of high-performance...', ' developing and using artificial intelligence (AI) to solve problems...']

Test 3: Unload and reload adapter
Output: [' great way to learn about the history of the world...', ' developing artificial intelligence systems that can be used to solve complex problems...']

================================================================================
SUCCESS: EMBEDDING_LORA_TESTS with lora_target_modules=['all'] PASSED!
================================================================================
```

Overall, the csgmv backend correctly handles `embed_tokens` and `lm_head` LoRA layers when using `--lora-target-modules=all`.

yushengsu-thu commented on test/srt/lora/test_lora_hf_sgl_logprob_diff.py:
```diff
@@ -543,6 +543,63 @@ def test_lora_logprob_comparison_full(self):
+
+        model_path = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"
+        lora_paths = ["ash256/sglang_embedding_lora_test_adapter"]
+        prompts = DEFAULT_TEST_PROMPTS[:2]
```
There are only 5 prompts. Just test all of them and do not need `DEFAULT_TEST_PROMPTS[:2]`

ashtonchew replied:
Fixed, now using all 5 prompts.

yushengsu-thu commented on test/srt/lora/test_lora_eviction.py:
```diff
@@ -73,6 +79,53 @@ def test_lora_eviction_with_reused_lora_name(self):
+                        )
+                    else:
+                        output_history[prompt] = output
+
```
To keep the codebase clean and easy to maintain, could you please simplify the code? If the test was generated using Cursor or another AI tool, it would be great if you could prune it and remove any redundant parts. 
Like following the format below:
```
def test_lora_eviction_with_embedding_lora_all_target_modules(self):
    self._run_test(....)
    self._run_test(....)
```

ashtonchew committed:
refactor(lora): simplify embedding LoRA tests per review feedback

- Use all DEFAULT_TEST_PROMPTS instead of [:2] slice in logprob diff tests
- Refactor test_lora_eviction_with_embedding_lora_all_target_modules to use _run_test()
- Add optional base_model, lora_target_modules, max_lora_rank params to _run_test()

ashtonchew commented on test/srt/lora/test_lora_hf_sgl_logprob_diff.py:
```diff
@@ -543,6 +543,63 @@ def test_lora_logprob_comparison_full(self):
+
+        model_path = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"
+        lora_paths = ["ash256/sglang_embedding_lora_test_adapter"]
+        prompts = DEFAULT_TEST_PROMPTS[:2]
```
Fixed, now using all 5 prompts.

Fridge003 commented on test/srt/lora/test_chunked_lora_embedding.py:
```diff
@@ -0,0 +1,96 @@
+"""Test ChunkedSgmvLoRABackend.run_lora_a_embedding() method."""
```
Please move this test to test/nightly

Fridge003 commented on python/sglang/srt/server_args.py:
```diff
@@ -4462,16 +4462,20 @@ def check_lora_server_args(self):
+                    # now supports embedding and lm_head layers via run_lora_a_embedding().
+                    # See: https://github.com/sgl-project/sglang/pull/14796
+                    #
+                    # # When using the chunked SGMV backend, skip embedding / lm_head layers for now,
```
We can remove these comment lines

ashtonchew replied:
Removed in latest revision.

Fridge003 commented on python/sglang/srt/lora/backend/chunked_backend.py:
```diff
@@ -254,6 +301,63 @@ def prepare_lora_batch(
         self.batch_info = batch_info
 
+        # Setup embedding_batch_info (uses original sequence structure, not chunked)
+        bs = forward_batch.batch_size
```
The same, can we skip this part when there is no embedding layers in lora adaptors?

ashtonchew replied:
Resolved as above.

ashtonchew committed:
refactor(lora): optimize embedding batch info for chunked backend

Skip embedding_batch_info creation when target_modules doesn't include
embed_tokens or lm_head, reducing unnecessary allocations.

- Add has_embedding_layers flag to conditionally initialize embedding
  batch info only when embedding LoRA layers are enabled
- Remove obsolete comments about csgmv embedding limitations
- Move chunked embedding test to nightly suite
- Restore lora_target_modules=["all"] in update tests

ashtonchew commented:
@Fridge003 here

ashtonchew commented:
@Fridge003 following up on this
